rfcs/context_uuid.md

```diff
@@ -0,0 +1,114 @@
+top-level window to initiate an action; with some work this allows
+supporting any case where it's possible to get a js handle between the
+test window and the top-level window. However we have no support for
+noopener cases or COOP headers, or other cases where it's not possible
+to pass messages between windows using js alone.
+
+To address these use cases, we propose allowing the use of a `uuid`
+query parameter on URLs, which will provide an alternate identifier
+for the window that loads the URL.
+
+In addition we propose updating the `__wptrunner_id` property name to
+`testdriver_id`, and making it a documented part of the harness
+setup. This will allow the default implementation of testdriver.js to
+refer to the property directly rather than it being part of the
+wptrunner implementation.
+
+testdriver will get a new function `get_context_id(ctx)`, with the
+following behaviour:
+* If `ctx` is a string, return `ctx`.
+* If ctx is a WindowProxy with a readable `location.href` which parses
+  as a URL and has a `uuid` query parameter, return the UUID parameter.
+* If ctx is a WindowProxy and it has a `testdriver_id` attribute,
+  return the value of the attribute, or if it doesn't have such an
+  attribute, but the Window is writable from the current context, set
+  `testdriver_id` to a new id and return that.
+* If `ctx` is a nested browsing context container with a `src`
+  attribute (for `iframe`, `frame`, `embed`) or `data` attribute (for
+  `object`) and the attribute value parses as a URL with a uuid
+  query parameter, return the uuid parameter.
+
+On the wptrunner side, we keep the same behaviour of searching through
+the tree of browsing contexts, but in addition we parse `uuid` query
+strings out of the attribute values of nested context containers and
+the URLs of loaded documents, and match those against the target URL.
+
+In this system a specific context may have more than one associated id
+(if the various ways of setting it differ), but we can use any
+associated id to get to the browsing context.
+
+For noopener cases, we can now provide a uuid in the URL and use that
+to invoke testdriver actions on the remote context:
+
+```
+// Use the server subn to generate a UUID
+let uuid = "{{uuid()}}"
+open(`child.html?uuid=${uuid}`)
+// [...] stuff to ensure child is loaded etc.
+// In this case the context identifier is used directly
+test_driver.delete_all_cookies(uuid)
+```
+
+We can also access a cross-origin-nested context:
+```
+<iframe src=https://{{host}}:{{ports[https][1]}}/file.html?uuid={{uuid()}}></iframe>
+<script>
+onload = () => {
+  let frame = frames[0];
+  test_driver.delete_all_cookies(frame)
+}
+</script>
+```
+
+TODO: How should this extend to other resources such as worker scripts. For
+current testdriver functionality it doesn't make much sense to run in
+a worker, but once we have cross-context messaging we might want to
+send messages to a worker we can't access via js.
```

Comment:
COEP tests contain script injection into workers/service workers (`executor_js_path` and `sw_executor_js_path` in `/wpt/html/cross-origin-embedder-policy/credentialless/resources/common.js`). I don't know the details though. FYI @ArthurSonzogni